allocation: clarify RestoreInProgressAllocationDecider failure message

schase-es opened the pull request:
The RestoreInProgressAllocationDecider can issue a grave message about shard restoration failure, when declining a shard allocation. Sometimes, this is because of restoration failure, but sometimes it is because another decider has declined the allocation. This change adds a check of UnassignedInfo to make this message appropriate.

Fixes ES-11809
Fixes #100233

schase-es committed:
allocation: clarify RestoreInProgressAllocationDecider failure message

The RestoreInProgressAllocationDecider can issue a grave message about shard
restoration failure, when declining a shard allocation. Sometimes, this is
because of restoration failure, but sometimes it is because another decider has
declined the allocation. This change adds a check of UnassignedInfo to make this
message appropriate.

Fixes ES-11809

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

schase-es commented:
I'm not sure about the first test case... it issues a message saying that the other allocation deciders did it. But there are no other allocation deciders...

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -49,15 +50,26 @@ public Decision canAllocate(final ShardRouting shardRouting, final RoutingAlloca
+                shardRouting.unassignedInfo().details()
+            );
+        } else {
+            return allocation.decision(
+                Decision.NO,
+                NAME,
+                "shard was prevented from being allocated on all nodes because of other allocation deciders"
+            );
```
I agree, I'm not 100% clear on why we should be returning NO if the specific thing we're responsible for is not in play here.

Am I right in understanding that the shard might have exhausted its retries because either
- There was some failure to restore the shard (e.g. file corruption)
- The shard could not be assigned anywhere (i.e. [this case](https://github.com/elastic/elasticsearch/blob/25e4a0600b63b44d83bf753c9b342c8822798ffd/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L969-L980))

And we want to distinguish the former from the latter? i.e. the `RestoreInProgressAllocationDecider` is right to say no, because we have stopped attempting to restore the shard, but there were no failed allocations so we shouldn't indicate that there will be failures in the logs.

I think there's a lot of assumed knowledge about the restore-in-progress lifecycle here and it would benefit from some comments or potentially restructuring.

DaveCTurner replied:
Ah I missed this conversation. Here's a test that shows the problem:

```diff
diff --git a/server/src/internalClusterTest/java/org/elasticsearch/snapshots/RestoreSnapshotIT.java b/server/src/internalClusterTest/java/org/elasticsearch/snapshots/RestoreSnapshotIT.java
index 953cddba0ab7..3741562deb40 100644
--- a/server/src/internalClusterTest/java/org/elasticsearch/snapshots/RestoreSnapshotIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/snapshots/RestoreSnapshotIT.java
@@ -12,6 +12,8 @@ package org.elasticsearch.snapshots;
 import org.apache.logging.log4j.Level;
 import org.elasticsearch.action.ActionFuture;
 import org.elasticsearch.action.ActionRequestBuilder;
+import org.elasticsearch.action.admin.cluster.allocation.ClusterAllocationExplainRequest;
+import org.elasticsearch.action.admin.cluster.allocation.TransportClusterAllocationExplainAction;
 import org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreSnapshotResponse;
 import org.elasticsearch.action.admin.indices.settings.get.GetSettingsResponse;
 import org.elasticsearch.action.admin.indices.template.get.GetIndexTemplatesResponse;
@@ -20,6 +22,8 @@ import org.elasticsearch.client.internal.Client;
 import org.elasticsearch.cluster.block.ClusterBlocks;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.ByteSizeUnit;
 import org.elasticsearch.core.TimeValue;
@@ -41,6 +45,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
@@ -1025,4 +1030,65 @@ public class RestoreSnapshotIT extends AbstractSnapshotIntegTestCase {
             mockLog.assertAllExpectationsMatched();
         }
     }
+
+    public void testExplainUnassigableDuringRestore() {
+        final String repoName = "repo-" + randomIdentifier();
+        createRepository(repoName, FsRepository.TYPE);
+        final String indexName = "index-" + randomIdentifier();
+        createIndexWithContent(indexName);
+        final String snapshotName = "snapshot-" + randomIdentifier();
+        createSnapshot(repoName, snapshotName, List.of(indexName));
+        assertAcked(indicesAdmin().prepareDelete(indexName));
+
+        final RestoreSnapshotResponse restoreSnapshotResponse = clusterAdmin().prepareRestoreSnapshot(
+            TEST_REQUEST_TIMEOUT,
+            repoName,
+            snapshotName
+        )
+            .setIndices(indexName)
+            .setRestoreGlobalState(false)
+            .setWaitForCompletion(true)
+            .setIndexSettings(
+                Settings.builder().put(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name", "not-a-node-" + randomIdentifier())
+            )
+            .get();
+
+        logger.info("--> restoreSnapshotResponse: {}", Strings.toString(restoreSnapshotResponse, true, true));
+        assertThat(restoreSnapshotResponse.getRestoreInfo().failedShards(), greaterThan(0));
+
+        final var clusterExplainResponse1 = client().execute(
+            TransportClusterAllocationExplainAction.TYPE,
+            new ClusterAllocationExplainRequest(TEST_REQUEST_TIMEOUT).setIndex(indexName).setShard(0).setPrimary(true)
+        ).actionGet();
+
+        logger.info("--> clusterExplainResponse1: {}", Strings.toString(clusterExplainResponse1, true, true));
+        for (var nodeDecision : clusterExplainResponse1.getExplanation()
+            .getShardAllocationDecision()
+            .getAllocateDecision()
+            .getNodeDecisions()) {
+            assertEquals(
+                Set.of("restore_in_progress", "filter"),
+                nodeDecision.getCanAllocateDecision().getDecisions().stream().map(Decision::label).collect(Collectors.toSet())
+            );
+        }
+
+        updateIndexSettings(Settings.builder().putNull(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name"), indexName);
+
+        final var clusterExplainResponse2 = client().execute(
+            TransportClusterAllocationExplainAction.TYPE,
+            new ClusterAllocationExplainRequest(TEST_REQUEST_TIMEOUT).setIndex(indexName).setShard(0).setPrimary(true)
+        ).actionGet();
+
+        logger.info("--> clusterExplainResponse2: {}", Strings.toString(clusterExplainResponse2, true, true));
+        for (var nodeDecision : clusterExplainResponse2.getExplanation()
+            .getShardAllocationDecision()
+            .getAllocateDecision()
+            .getNodeDecisions()) {
+            assertEquals(
+                Set.of("restore_in_progress"),
+                nodeDecision.getCanAllocateDecision().getDecisions().stream().map(Decision::label).collect(Collectors.toSet())
+            );
+        }
+
+    }
 }

```

Note that the shard couldn't be allocated because of the filter, but then when the filter is removed we still mustn't allocate the shard because the restore is no longer in progress.

DaveCTurner replied:
The message in this latter case is:

`shard has failed to be restored from the snapshot [default:repo-sdvofzvibks:snapshot-iqqndiwoes/HWiV1JnTSRKor0KieINqWA] - manually close or delete the index [index-djtuceqgi] in order to retry to restore the snapshot again or use the reroute API to force the allocation of an empty primary shard. Details: [restore_source[repo-sdvofzvibks/snapshot-iqqndiwoes]]`

The issue is that this message sends folks off trying to look for problems with the snapshot repository when in fact we never even tried to contact the snapshot repository, the obstacle was elsewhere, and may no longer exist. Moreover the `Details:` bit at the end is pointless, it's actually less detailed than the snapshot ID `default:repo-sdvofzvibks:snapshot-iqqndiwoes/HWiV1JnTSRKor0KieINqWA` logged earlier.

schase-es replied:
I think I understand now -- this is more of an issue with how a cluster state change that voids the context of the restore (the index underneath it changes) cascades through the allocator.

I'm seeing where this happens in RestoreService.updateRestoreStateWithDeletedIndices. And where failed/completed restores are cleaned out in RestoreService.removeCompletedRestoresFromClusterState.

schase-es replied:
It seems like it might be more accurate in the allocation failure message to state that some aspect of the cluster state context has changed.

schase-es committed:
Addressed review feedback:
- decider says yes for irrelevant cases
- tests are updated
- comment explains the state being checked at the end

schase-es committed:
Merge branch 'main' into ES-11809_restore-in-progress-allocation-failure-message

mhl-b reviewed (approved):
LGTM, thanks for thorough explanation, it helps. I would change last YES message. Don't need second review from me.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -49,14 +50,30 @@ public Decision canAllocate(final ShardRouting shardRouting, final RoutingAlloca
+
         return allocation.decision(
-            Decision.NO,
+            Decision.YES,
```
This changes the behaviour of the decider and I don't think we should do that. It's important that we return `NO` if the restore UUID isn't in the cluster state any more. We just need to explain that in this case the restore didn't fail, we just never even attempted it before the restore concluded.

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -49,14 +50,30 @@ public Decision canAllocate(final ShardRouting shardRouting, final RoutingAlloca
+
+        /**
+         * POST: the RestoreInProgress.ShardRestoreStatus is either failed or succeeded. This section turns a
+         * turn a shard failure into a NO decision to allocate. See {@link AllocationService.applyFailedShards}
```
"turns a turn a shard..." typo in javadoc?

schase-es committed:
Reverting commit re: David's feedback

schase-es committed:
Adding David's test for differentiating between allocation failure and restore faults.

schase-es committed:
Minor detail: fixing up a test error message

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -49,15 +50,31 @@ public Decision canAllocate(final ShardRouting shardRouting, final RoutingAlloca
+            return allocation.decision(
+                Decision.NO,
+                NAME,
+                "shard was prevented from being allocated on all nodes because of other allocation deciders in previous rounds"
```
Could this be something like
"Restore from snapshot failed because the configured constraints prevented allocation on any of the available nodes. Please check constraints applied in index and cluster settings, then retry the restore"

Just to avoid the talk about other allocation deciders (that seems like an abstract concept, but I might be misunderstanding who the audience is here) and also the tip to check the constraints and retry might be helpful?

Happy to go with consensus here

schase-es replied:
I'm sure -- maybe we can create a separate task for later to pend the reasoning in RestoreInProgress to somewhere in UnassignedInfo/ClusterState

nicktindall reviewed (approved):
This LGTM, pending clarification on the message

schase-es committed:
Merge branch 'main' into ES-11809_restore-in-progress-allocation-failure-message

schase-es committed:
Fixing up error message, and testing for it in RestoreSnapshotIT

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
Merge branch 'main' into ES-11809_restore-in-progress-allocation-failure-message

schase-es committed:
Merge branch 'main' into ES-11809_restore-in-progress-allocation-failure-message

schase-es committed:
Style fixes

schase-es committed:
Merge branch 'main' into ES-11809_restore-in-progress-allocation-failure-message
